cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -91,51 +193,601 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+    implements NotificationHandler<StreamSizeNotification> {
+    // Key is the schedule ID
+    private final ConcurrentMap<String, StreamSizeScheduleTask> scheduleTasks;
+    private final Id.Stream streamId;
+    private final AtomicInteger activeTasks;
+
+    private Cancellable notificationSubscription;
+    private StreamSizeNotification lastNotification;
+    private StreamSize lastPollingInfo;
+    private Long delta;
+
+    private StreamSubscriber(Id.Stream streamId) {
+      this.streamId = streamId;
+      this.scheduleTasks = Maps.newConcurrentMap();
+      this.activeTasks = new AtomicInteger(0);
+      this.delta = null;
+    }
+
+    @Override
+    protected void startUp() throws Exception {
+      notificationSubscription = notificationService.subscribe(getFeed(), this, sendPollingInfoExecutor);
+    }
+
+    @Override
+    protected void shutDown() throws Exception {
+      if (notificationSubscription != null) {
+        notificationSubscription.cancel();
+      }
+    }
+
+    @Override
+    protected void runOneIteration() throws Exception {
+      if (activeTasks.get() == 0) {
+        return;
+      }
+
+      try {
+        StreamSize streamSize = pollOnce();
+        sendPollingInfoToActiveTasks(new StreamSizeNotification(streamSize.getTimestamp(), streamSize.getSize()));
+      } catch (Throwable t) {
+        LOG.error("Could not poll stream {}", streamId.getName(), t);
+      }
+    }
+
+    @Override
+    protected Scheduler scheduler() {
+      return Scheduler.newFixedRateSchedule(pollingDelay, pollingDelay, TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    protected ScheduledExecutorService executor() {
+      return streamPollingExecutor;
+    }
+
+    @Override
+    public Type getNotificationFeedType() {
+      return StreamSizeNotification.class;
+    }
+
+    @Override
+    public void received(final StreamSizeNotification notification, NotificationContext notificationContext) {
+      // We only use the stream size notification to the schedule tasks if it came after the last seen notification
+      boolean send = false;
+      if (lastNotification == null || notification.getTimestamp() > lastNotification.getTimestamp()) {
+        send = true;
+        lastNotification = notification;
+      }
+      if (send) {
+        if (activeTasks.get() > 0) {
+
+          boolean poll = false;
+          if (delta == null) {
+            poll = true;
+          } else {
+            for (StreamSizeScheduleTask streamSizeScheduleTask : scheduleTasks.values()) {
+              if (streamSizeScheduleTask.shouldTriggerProgram(notification.getSize() - delta)) {
+                poll = true;
+                break;
+              }
+            }
+          }
+
+          if (poll) {
+            pollBookingExecutor.schedule(new Runnable() {
+              @Override
+              public void run() {
+                try {
+                  StreamSize streamSize = pollOnce();
```

Comment:
Yes it can happen, and I think it is safe too. pollOnce() is synchronized, and so the task.reset method won't be called for wrong reasons. It might lead to task.receivedPollingInformation() called concurrently, and the synchronized block in it will make sure that the program is not run twice.